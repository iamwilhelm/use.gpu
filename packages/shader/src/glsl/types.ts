import {
  ParsedModuleCache as ParsedModuleCacheT,
  ParsedBundle as ParsedBundleT,
  ParsedModule as ParsedModuleT,
  ShaderModule as ShaderModuleT,
  VirtualTable as VirtualTableT,
  DataBinding as DataBindingT,
} from '../types';

export type {
  CompressedNode,
  ShakeTable,
  ShakeOp,
  VirtualRender,
  StorageSource,
  UniformAttribute,
  UniformAttributeValue,
} from '../types';

export type ParsedModuleCache = ParsedModuleCacheT<SymbolTable>;
export type ParsedBundle = ParsedBundleT<SymbolTable>;
export type ParsedModule = ParsedModuleT<SymbolTable>;
export type ShaderModule = ShaderModuleT<SymbolTable>;
export type VirtualTable = VirtualTableT<SymbolTable>;
export type DataBinding = DataBindingT<SymbolTable>;

export type ShaderCompiler = (code: string, stage: string) => Uint8Array | Uint32Array;

export type ComboRef = ModuleRef | FunctionRef | DeclarationRef;

export type SymbolTable = {
  hash: string,
  symbols?: SymbolRef[],
  visibles?: SymbolRef[],
  globals?: SymbolRef[],
  modules?: ModuleRef[],
  functions?: FunctionRef[],
  declarations?: DeclarationRef[],
  externals?: DeclarationRef[],
};

export enum RefFlags {
  Exported = 1,
  Optional = 1 << 1,
  Global   = 1 << 2,
};

export type SymbolRef = string;

export type ImportRef = {
  name: string,
  imported: string,
}

export type SymbolsRef = {
  at: number,
  symbols: SymbolRef[],
}

export type ModuleRef = SymbolsRef & {
  name: string,
  imports: ImportRef[],
}

export type FunctionRef = SymbolsRef & {
  prototype: PrototypeRef,
  identifiers?: string[],
  flags: RefFlags,
}

export type DeclarationRef = SymbolsRef & {
  prototype?: PrototypeRef,
  variable?: VariableRef,
  struct?: QualifiedStructRef,
  identifiers?: string[],
  flags: RefFlags,
}

export type TypeRef = {
  name: string,
  qualifiers?: string[],
  members?: MemberRef[],
}

export type PrototypeRef = {
  type: TypeRef,
  name: string,
  parameters: string[],
}

export type VariableRef = {
  type: TypeRef,
  locals: LocalRef[],
}

export type MemberRef = {
  type: TypeRef,
  name: string,
}

export type LocalRef = {
  name: string,
  expr: any,
}

export type QualifiedStructRef = {
  type: TypeRef,
  name: string,
  struct: StructRef,
}

export type StructRef = {
  members: MemberRef[],
}

export type ShaderDefine = string | number | boolean | null | undefined;
