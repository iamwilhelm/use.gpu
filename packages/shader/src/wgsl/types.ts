import {
  ParsedModuleCache as ParsedModuleCacheT,
  ParsedBundle as ParsedBundleT,
  ParsedModule as ParsedModuleT,
  ShaderModule as ShaderModuleT,
  VirtualTable as VirtualTableT,
  DataBinding as DataBindingT,
  ImportRef,
  RefFlags,
} from '../types';

export {
  RefFlags,
} from '../types';

export type {
  CompressedNode,
  ImportRef,
  ShakeTable,
  ShakeOp,
  ShaderDefine,
  StorageSource,
  UniformAttribute,
  UniformAttributeValue,
  VirtualRender,
} from '../types';

export type ParsedModuleCache = ParsedModuleCacheT<SymbolTable>;
export type ParsedBundle = ParsedBundleT<SymbolTable>;
export type ParsedModule = ParsedModuleT<SymbolTable>;
export type ShaderModule = ShaderModuleT<SymbolTable>;
export type VirtualTable = VirtualTableT<SymbolTable>;
export type DataBinding = DataBindingT<SymbolTable>;

export type SymbolTable = {
  hash: string,
  modules?: ModuleRef[],
  declarations?: DeclarationRef[],
  externals?: DeclarationRef[],
  symbols?: string[],
  visibles?: string[],
  globals?: string[],
};

export type ModuleRef = {
  at: number,
  symbols: string[],
  name: string,
  imports: ImportRef[],
};

export type DeclarationRef = {
  at: number,
  symbol: string,
  flags: RefFlags,
  func?: FunctionRef,
  variable?: VariableRef,
  constant?: VariableRef,
  alias?: TypeAliasRef,
  struct?: StructRef,
};

export type AttributesRef = {
  attributes?: AttributeRef[],
};

export type IdentifiersRef = {
  identifiers?: string[],
};

export type AttributeRef = {
  name: string,
  args?: string[],
};

export type TypeRef = {
  name: string,
  args?: TypeRef[],
};

export type TypeAliasRef = AttributesRef & {
  name: string,
  type: TypeRef,
};

export type QualifiedTypeAliasRef = TypeAliasRef & {
  qual?: string,
};

export type FunctionRef = AttributesRef & IdentifiersRef & FunctionHeaderRef;
export type VariableRef = AttributesRef & IdentifiersRef & VariableDeclarationRef;
export type AnnotatedTypeRef = AttributesRef & TypeRef;

export type ParameterRef = AttributesRef & {
  name: string,
  type: TypeRef,
};

export type FunctionHeaderRef = {
  name: string,
  type: AnnotatedTypeRef,
  parameters?: ParameterRef[] | string[],
};

export type VariableDeclarationRef = {
  name: string,
  type: TypeRef,
  qual?: string,
  value?: string,
};

export type StructMemberRef = AttributesRef & {
  name: string,
  type: TypeRef,
};

export type StructRef = AttributesRef & {
  name: string,
  members: StructMemberRef[],
};
