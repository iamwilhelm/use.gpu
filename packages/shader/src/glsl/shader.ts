import { Tree } from '@lezer/common';
import { ParsedModule, ParsedModuleCache, ShaderDefine } from './types';

import { makeLoadModule, makeLoadModuleWithCache } from '../util/shader';
import { makeBundleToAttribute, makeBundleToAttributes } from '../util/bundle';
import { decompressAST } from '../util/tree';

import { makeASTParser, compressAST } from './ast';
import { toTypeString, toTypeArgs } from './type';
import { parser } from './grammar/glsl';
import LRU from 'lru-cache';

export { loadStaticModule, loadVirtualModule } from '../util/shader';

// LRU cache for parsed shader code
export const makeModuleCache = (options: Record<string, any> = {}) => new LRU<string, ParsedModule>({
  max: 100,
  ...options,
});

export const DEFAULT_CACHE = makeModuleCache();

// Parse GLSL code into lezer tree
export const parseShader = (code: string): Tree => parser.parse(code);

// Parse a code module into its in-memory representation
// (AST + symbol table)
export const loadModule = makeLoadModule(parseShader, makeASTParser, compressAST);

// Use cache to load modules
export const loadModuleWithCache = makeLoadModuleWithCache(loadModule, DEFAULT_CACHE);

// Make GLSL definitions
export const defineConstants = (defs: Record<string, ShaderDefine>): string => {
  const out = [];
  for (let k in defs) if (defs[k] !== false && defs[k] !== null) out.push(`#define ${k} ${defs[k]}`);
  return out.join("\n");
}

export const bundleToAttribute = makeBundleToAttribute(toTypeString, toTypeArgs);
export const bundleToAttributes = makeBundleToAttributes(toTypeString, toTypeArgs);
