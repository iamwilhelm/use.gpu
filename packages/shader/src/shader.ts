import { Tree, SyntaxNode } from '@lezer/common';
import { parser } from './grammar/glsl';

export const parseGLSL = (code: string): Tree => {
  const parsed = parser.parse(code);
  return parsed;
}
