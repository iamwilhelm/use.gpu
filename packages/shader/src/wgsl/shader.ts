import { Tree } from '@lezer/common';
import { ParsedModule, ShaderDefine } from './types';

import { makeLoadModule, makeLoadModuleWithCache } from '../util/shader';
import { makeBundleToAttribute, makeBundleToAttributes } from '../util/bundle';
import { makeTranspile } from '../util/transpile';

import { makeASTParser, compressAST, decompressAST } from './ast';
import { toTypeSymbol, toTypeArgs } from './type';
import { removeComments, removeWhiteSpace, renameLocals } from './minify';
import { parser } from './grammar/wgsl';

import LRU from 'lru-cache';
import zip from 'lodash/zip';

export { loadStaticModule, loadVirtualModule, bindEntryPoint } from '../util/shader';

/** LRU cache for parsed shader code */
export const makeModuleCache = (options: Record<string, any> = {}) => new LRU<number, ParsedModule>({
  max: 100,
  ...options,
});

export const DEFAULT_CACHE = makeModuleCache();

/** Parse WGSL code into lezer tree */
export const parseShader = (code: string): Tree => parser.parse(code);

/** Parse a code module into its in-memory representation (AST + symbol table) */
export const loadModule = makeLoadModule(parseShader, makeASTParser, compressAST, decompressAST);

/** Use cache to load modules */
export const loadModuleWithCache = makeLoadModuleWithCache(loadModule, DEFAULT_CACHE);

/** Make WGSL constant definitions */
export const defineConstants = (defs: Record<string, ShaderDefine>): string => {
  const out = [];
  for (const k in defs) if (k[0] !== '@' && defs[k] != null) out.push(`const ${k} = ${defs[k]};`);
  return out.join("\n");
}

/** Make WGSL constant definitions */
export const defineEnables = (enabled: string[]) => enabled.length ? `enable ${enabled.join(', ')};` : '';

/** Convert a bundle with a defined entry point to a definition for that attribute or type. */
export const bundleToAttribute = makeBundleToAttribute(toTypeSymbol, toTypeArgs);

/** Convert a bundle to a definition for all its attributes. */
export const bundleToAttributes = makeBundleToAttributes(toTypeSymbol, toTypeArgs);

// Simple whitespace / comment removal
const minifyCode = (code: string) => {
  code = removeComments(code);
  code = renameLocals(code);
  code = removeWhiteSpace(code);
  return code;
};

export const symbolDictionary = {
  A: 'at' as 'at',
  B: 'bindings' as 'bindings',
  C: 'vec4<f32>' as 'vec4<f32>',
  D: 'vec3<f32>' as 'vec3<f32>',
  E: 'exports' as 'exports',
  F: 'func' as 'func',
  G: 'flags' as 'flags',
  H: 'inferred' as 'inferred',
  I: 'identifiers' as 'identifiers',
  J: 'imported' as 'imported',
  K: 'imports' as 'imports',
  L: 'linkable' as 'linkable',
  M: 'members' as 'members',
  N: 'name' as 'name',
  O: 'modules' as 'modules',
  P: 'parameters' as 'parameters',
  Q: 'qual' as 'qual',
  R: 'symbol' as 'symbol',
  S: 'symbols' as 'symbols',
  T: 'type' as 'type',
  U: 'struct' as 'struct',
  V: 'variable' as 'variable',
  W: 'visibles' as 'visibles',
  X: 'externals' as 'externals',
  Y: 'types' as 'types',
  Z: 'attr' as 'attr',
};

/** ES/CommonJS Transpiler */
export const transpileWGSL = makeTranspile('wgsl', 'wgsl', symbolDictionary, loadModule, compressAST, minifyCode);

/** Templated literal syntax:

```tsx
wgsl`...`
``` */
export const wgsl = (literals: TemplateStringsArray, ...tokens: string[]) => {
  const code = zip(literals, tokens).flat();
  return loadModuleWithCache(code.join(''));
};

/** Format `number` as WGSL `f32` */
export const f32 = (x: number) => {
  const s = x.toString();
  return (!s.match(/\./)) ? s + '.0' : s;
};
/** Format `number` as WGSL `u32` */
export const u32 = (x: number) => Math.round(x).toString() + 'u';
/** Format `number` as WGSL `i32` */
export const i32 = (x: number) => Math.round(x).toString();
