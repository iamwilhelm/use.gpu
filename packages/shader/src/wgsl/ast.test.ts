//import { GLSLModules } from '@use-gpu/glsl';
import { parseShader } from './shader';
import { makeASTParser /*, rewriteUsingAST, resolveShakeOps, compressAST, decompressAST*/ } from './ast';
import { formatAST, hasErrorNode } from '../util/tree';
import { addASTSerializer } from '../test/snapshot';

addASTSerializer(expect);

describe('ast', () => {
  
  const makeGuardedParser = (code: any, tree: any): ReturnType<typeof makeASTParser> => {
    let errorNode = hasErrorNode(tree);
    if (errorNode) {
      for (let i = 0; i < 2; ++i) if (errorNode.parent) errorNode = errorNode.parent;
      console.log(formatAST(errorNode, code));
      throw new Error("Error in AST");
    }
    
    return makeASTParser(code, tree);
  }

  it('gets test imports', () => {
    const code = `
    import {MeshVertex} from 'use/types';
    import {viewUniforms as view, worldToClip} from 'use/view';
    import {getQuadUV} from 'geometry/quad';

    import 'test';

    @optional @link fn getInt() -> i32 {}

    @export fn main() {}
    `;

    const tree = parseShader(code);
    const {getImports} = makeGuardedParser(code, tree);

    const imports = getImports();
    expect(imports).toMatchSnapshot();
  });

  it('gets test var/const declarations', () => {
    const code = `
      var x: f32;
      var y: f32;
      let a: i32 = 3;
      type integer = i32;
      override b: i32;
    `;

    const tree = parseShader(code);
    const {getDeclarations} = makeGuardedParser(code, tree);

    const declarations = getDeclarations();
    expect(declarations).toMatchSnapshot();
  });

  it('gets test empty function declaration', () => {
    const code = `
      @export fn main() {}
    `;

    const tree = parseShader(code);
    const {getDeclarations} = makeGuardedParser(code, tree);

    const declarations = getDeclarations();
    expect(declarations).toMatchSnapshot();
  });

  it('gets test function declaration', () => {
    const code = `
      @fragment fn fragShader(in1: A, @location(2) in2: f32) -> @location(0) vec4<f32> {
        return foo(in1, in2);
      }
    `;

    const tree = parseShader(code);
    const {getDeclarations} = makeGuardedParser(code, tree);

    const declarations = getDeclarations();
    expect(declarations).toMatchSnapshot();
  });

  it('gets test struct declaration', () => {
    const code = `
      struct light {
        intensity: f32,
        @annotate position: vec3<f32>,
      }
    `;

    const tree = parseShader(code);
    const {getDeclarations} = makeGuardedParser(code, tree);

    const declarations = getDeclarations();
    expect(declarations).toMatchSnapshot();
  });

  it('gets test declarations with array', () => {
    const code = `
      let QUAD: array<vec2<i32>, 4> = array<vec2<i32>, 4>(
        vec2<i32>(0, 0),
        vec2<i32>(1, 0),
        vec2<i32>(0, 1),
        vec2<i32>(1, 1),
      );
    `;

    const tree = parseShader(code);
    const {getDeclarations} = makeGuardedParser(code, tree);

    const declarations = getDeclarations();
    expect(declarations).toMatchSnapshot();
  });

  it('gets symbol table', () => {
    const code = `
      @exported var x: f32;
      var y: f32;
      let a: i32 = 3;
      type integer = i32;
      override b: i32;

      @exported struct light {
        intensity: f32,
        @annotate position: vec3<f32>,
      }

      let QUAD: array<vec2<i32>, 4> = array<vec2<i32>, 4>(
        vec2<i32>(0, 0),
        vec2<i32>(1, 0),
        vec2<i32>(0, 1),
        vec2<i32>(1, 1),
      );

      @optional @link fn getInt() -> i32 {}

      @export fn main() {}
    `;

    const tree = parseShader(code);
    const {getSymbolTable} = makeGuardedParser(code, tree);

    const table = getSymbolTable();
    expect(table).toMatchSnapshot();
  });

  it('gets shake table', () => {
    const code = `
      @exported var x: f32;
      var y: f32;

      @optional @link fn getFloat1() -> f32 {}

      @optional @link fn getFloat2() -> f32 { return x + y; }

      @export fn main() {
        var z: f32 = getFloat1() + getFloat2();
      }
    `;

    const tree = parseShader(code);
    const {getShakeTable} = makeGuardedParser(code, tree);

    const table = getShakeTable();
    expect(table).toMatchSnapshot();
  });

  /*
  it('gets quad vertex imports', () => {
    const code = GLSLModules['instance/vertex/quad'];

    const tree = parseShader(code);
    const {getImports} = makeGuardedParser(code, tree);

    const imports = getImports();
    expect(imports).toMatchSnapshot();
  });
  
  it('gets quad vertex functions', () => {
    const code = GLSLModules['instance/vertex/quad'];

    const tree = parseShader(code);
    const {getFunctions} = makeGuardedParser(code, tree);

    const functions = getFunctions();
    expect(functions).toMatchSnapshot();
  });

  it('gets quad vertex declarations', () => {
    const code = GLSLModules['instance/vertex/quad'];

    const tree = parseShader(code);
    const {getDeclarations} = makeGuardedParser(code, tree);

    const declarations = getDeclarations();
    expect(declarations).toMatchSnapshot();
  });

  it('gets solid fragment declarations', () => {
    const code = GLSLModules['instance/fragment/solid'];

    const tree = parseShader(code);
    const {getDeclarations} = makeGuardedParser(code, tree);

    const declarations = getDeclarations();
    expect(declarations).toMatchSnapshot();
  });

  it('gets use view declarations', () => {
    const code = GLSLModules['use/view'];

    const tree = parseShader(code);
    const {getDeclarations} = makeGuardedParser(code, tree);

    const declarations = getDeclarations();
    expect(declarations).toMatchSnapshot();
  });

  it('gets quad vertex symbol table', () => {
    const code = GLSLModules['instance/vertex/quad'];

    const tree = parseShader(code);
    const {getSymbolTable} = makeGuardedParser(code, tree);

    const symbolTable = getSymbolTable();
    expect(symbolTable).toMatchSnapshot();
  });

  it('gets geometry quad symbol table', () => {
    const code = GLSLModules['geometry/quad'];

    const tree = parseShader(code);
    const {getSymbolTable} = makeGuardedParser(code, tree);

    const symbolTable = getSymbolTable();
    expect(symbolTable).toMatchSnapshot();
  });

  it('gets use view symbol table', () => {
    const code = GLSLModules['use/view'];

    const tree = parseShader(code);
    const {getSymbolTable} = makeGuardedParser(code, tree);

    const symbolTable = getSymbolTable();
    expect(symbolTable).toMatchSnapshot();
  });

  it('gets use types symbol table', () => {
    const code = GLSLModules['use/types'];

    const tree = parseShader(code);
    const {getSymbolTable} = makeGuardedParser(code, tree);

    const symbolTable = getSymbolTable();
    expect(symbolTable).toMatchSnapshot();
  });

  it('rewrites code using the AST', () => {
    const code = `
    float getValue(int index);
    void main() {
      float x = 3.0;
      float y = getValue(2);
    }
    `;

    const tree = parseShader(code);
    const rename = new Map<string, string>();
    rename.set('main', 'entryPoint');
    rename.set('getValue', '_zz_getValue');
    
    const output = rewriteUsingAST(code, tree, rename);
    expect(output).toMatchSnapshot();
  });

  it('rewrites code using the compressed AST', () => {
    const code = `
    float getValue(int index);
    void main() {
      float x = 3.0;
      float y = getValue(2);
      vec3 v;
      v.xyz;
    }
    `;

    const tree = parseShader(code);
    const rename = new Map<string, string>();
    rename.set('main', 'entryPoint');
    rename.set('getValue', '_zz_getValue');
    
    const compressed = compressAST(tree);
    const decompressed = decompressAST(compressed);
    expect(compressed).toMatchSnapshot();
    expect(decompressed).toMatchSnapshot();

    const output1 = rewriteUsingAST(code, tree, rename);
    const output2 = rewriteUsingAST(code, decompressed, rename);
    expect(output2).toEqual(output1);
  });

  it('rewrites a lot of code using the compressed AST', () => {
    const code = `
#pragma import {SolidVertex} from 'use/types'

SolidVertex getVertex(int, int);

#ifdef IS_PICKING
layout(location = 0) out flat uint fragIndex;
#else
layout(location = 0) out vec4 fragColor;
layout(location = 1) out vec2 fragUV;
#endif

void main() {
  int vertexIndex = gl_VertexIndex;
  int instanceIndex = gl_InstanceIndex;

  SolidVertex v = getVertex(vertexIndex, instanceIndex);

  gl_Position = v.position;
#ifdef IS_PICKING
  fragIndex = uint(instanceIndex);
#else
  fragColor = v.color;
  fragUV = v.uv;
#endif
}
    `;

    const tree = parseShader(code);
    const rename = new Map<string, string>();
    rename.set('main', 'entryPoint');
    rename.set('getVertex', '_zz_getVertex');
    
    const compressed = compressAST(tree);
    const decompressed = decompressAST(compressed);
    expect(compressed).toMatchSnapshot();
    expect(decompressed).toMatchSnapshot();

    const output1 = rewriteUsingAST(code, tree, rename);
    const output2 = rewriteUsingAST(code, decompressed, rename);
    expect(output2).toEqual(output1);
  });

  it('recompresses AST', () => {
    const code = `
#pragma import {SolidVertex} from 'use/types'

SolidVertex getVertex(int, int);

#ifdef IS_PICKING
layout(location = 0) out flat uint fragIndex;
#else
layout(location = 0) out vec4 fragColor;
layout(location = 1) out vec2 fragUV;
#endif

void main() {
  int vertexIndex = gl_VertexIndex;
  int instanceIndex = gl_InstanceIndex;

  SolidVertex v = getVertex(vertexIndex, instanceIndex);

  gl_Position = v.position;
#ifdef IS_PICKING
  fragIndex = uint(instanceIndex);
#else
  fragColor = v.color;
  fragUV = v.uv;
#endif
}
    `;

    const tree = parseShader(code);
    const compressed = compressAST(tree);
    const decompressed = decompressAST(compressed);
    const recompressed = compressAST(decompressed);
    expect(compressed).toEqual(recompressed);
  });

  it('shakes simple program', () => {
    const code = `
const float x = 1.0;

#pragma export
float getA() {
  return x;
}

#pragma export
float getB() {
  return x;
}
    `;

    const tree = parseShader(code);
    const table = makeGuardedParser(code, tree).getShakeTable();

    expect(table).toBeTruthy();
    expect(table).toMatchSnapshot();
    if (!table) return;
    
    const keep = new Set(['getA']);
    const ops = resolveShakeOps(table, keep);
    expect(rewriteUsingAST(code, tree, new Map(), ops)).toMatchSnapshot();
  });
  
  it('shakes use/view AST', () => {
    const code = GLSLModules['use/view'];

    const tree = parseShader(code);
    const table = makeGuardedParser(code, tree).getShakeTable();

    expect(table).toBeTruthy();
    expect(table).toMatchSnapshot();
    if (!table) return;
    
    const keep = new Set(['worldToClip']);
    const ops = resolveShakeOps(table, keep);
    expect(rewriteUsingAST(code, tree, new Map(), ops)).toMatchSnapshot();
  });

  it('shakes use/view AST using compressed AST', () => {
    const code = GLSLModules['use/view'];

    const tree = parseShader(code);
    const table = makeGuardedParser(code, tree).getShakeTable();

    expect(table).toBeTruthy();
    expect(table).toMatchSnapshot();
    if (!table) return;
    
    const keep = new Set(['worldToClip']);
    const ops = resolveShakeOps(table, keep);
    
    const tree1 = tree;
    const tree2 = decompressAST(compressAST(tree1));

    const code1 = rewriteUsingAST(code, tree1, new Map(), ops);
    const code2 = rewriteUsingAST(code, tree2, new Map(), ops);

    expect(code2).toEqual(code1);
  });

  it('gets shake information for instance/vertex/quad AST', () => {
    const code = GLSLModules['instance/vertex/quad'];

    const tree = parseShader(code);
    const table = makeGuardedParser(code, tree).getShakeTable();
    expect(table).toMatchSnapshot();
    
  });

  it('gets shake information for geometry/quad AST', () => {
    const code = GLSLModules['geometry/quad'];

    const tree = parseShader(code);
    const table = makeGuardedParser(code, tree).getShakeTable();
    expect(table).toMatchSnapshot();
    
  });

  it('gets shake information for use/types AST', () => {
    const code = GLSLModules['use/types'];

    const tree = parseShader(code);
    const table = makeGuardedParser(code, tree).getShakeTable();
    expect(table).toMatchSnapshot();
    
  });
  */
});

