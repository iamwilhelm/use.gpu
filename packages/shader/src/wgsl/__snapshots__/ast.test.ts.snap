// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ast gets shake table 1`] = `
Array [
  Array [
    6,
    Array [
      "x",
      "getFloat2",
      "main",
    ],
  ],
  Array [
    27,
    Array [
      "y",
      "getFloat2",
      "main",
    ],
  ],
  Array [
    49,
    Array [
      "getFloat1",
      "main",
    ],
  ],
  Array [
    100,
    Array [
      "getFloat2",
      "main",
    ],
  ],
  Array [
    166,
    Array [
      "main",
    ],
  ],
]
`;

exports[`ast gets symbol table 1`] = `
Object {
  "declarations": Array [
    Object {
      "at": 6,
      "flags": 1,
      "symbol": "x",
      "variable": Object {
        "attributes": Array [
          Object {
            "args": undefined,
            "name": "exported",
          },
        ],
        "identifiers": Array [],
        "name": "x",
        "qual": undefined,
        "type": Object {
          "args": undefined,
          "name": "f32",
        },
        "value": undefined,
      },
    },
    Object {
      "at": 27,
      "flags": 0,
      "symbol": "y",
      "variable": Object {
        "attributes": undefined,
        "identifiers": Array [],
        "name": "y",
        "qual": undefined,
        "type": Object {
          "args": undefined,
          "name": "f32",
        },
        "value": undefined,
      },
    },
    Object {
      "at": 44,
      "constant": Object {
        "attributes": undefined,
        "identifiers": Array [],
        "name": "a",
        "type": Object {
          "args": undefined,
          "name": "i32",
        },
        "value": "3",
      },
      "flags": 0,
      "symbol": "a",
    },
    Object {
      "alias": Object {
        "attributes": undefined,
        "name": "integer",
        "type": Object {
          "args": undefined,
          "name": "i32",
        },
      },
      "at": 65,
      "flags": 0,
      "symbol": "integer",
    },
    Object {
      "at": 88,
      "constant": Object {
        "attributes": undefined,
        "identifiers": Array [],
        "name": "b",
        "type": Object {
          "args": undefined,
          "name": "i32",
        },
        "value": undefined,
      },
      "flags": 0,
      "symbol": "b",
    },
    Object {
      "at": 115,
      "flags": 1,
      "struct": Object {
        "attributes": Array [
          Object {
            "args": undefined,
            "name": "exported",
          },
        ],
        "members": Array [
          Object {
            "attributes": undefined,
            "name": "intensity",
            "type": Object {
              "args": undefined,
              "name": "f32",
            },
          },
          Object {
            "attributes": Array [
              Object {
                "args": undefined,
                "name": "annotate",
              },
            ],
            "name": "position",
            "type": Object {
              "args": Array [
                Object {
                  "args": undefined,
                  "name": "f32",
                },
              ],
              "name": "vec3",
            },
          },
        ],
        "name": "light",
      },
      "symbol": "light",
    },
    Object {
      "at": 199,
      "constant": Object {
        "attributes": undefined,
        "identifiers": Array [],
        "name": "QUAD",
        "type": Object {
          "args": Array [
            Object {
              "args": Array [
                Object {
                  "args": undefined,
                  "name": "i32",
                },
              ],
              "name": "vec2",
            },
            Object {
              "name": "4",
            },
          ],
          "name": "array",
        },
        "value": "array<vec2<i32>, 4>(
        vec2<i32>(0, 0),
        vec2<i32>(1, 0),
        vec2<i32>(0, 1),
        vec2<i32>(1, 1),
			)",
      },
      "flags": 0,
      "symbol": "QUAD",
    },
    Object {
      "at": 372,
      "flags": 6,
      "func": Object {
        "attributes": Array [
          Object {
            "args": undefined,
            "name": "optional",
          },
          Object {
            "args": undefined,
            "name": "external",
          },
        ],
        "identifiers": Array [],
        "name": "getInt",
        "parameters": undefined,
        "type": Object {
          "args": undefined,
          "attributes": undefined,
          "name": "i32",
        },
      },
      "symbol": "getInt",
    },
    Object {
      "at": 420,
      "flags": 0,
      "func": Object {
        "attributes": Array [
          Object {
            "args": undefined,
            "name": "export",
          },
        ],
        "identifiers": Array [],
        "name": "main",
        "parameters": undefined,
        "type": Object {
          "attributes": undefined,
          "name": "void",
        },
      },
      "symbol": "main",
    },
  ],
  "globals": undefined,
  "hash": "mtyot4jkqo",
  "modules": undefined,
  "symbols": Array [
    "x",
    "y",
    "a",
    "integer",
    "b",
    "light",
    "QUAD",
    "getInt",
    "main",
  ],
  "visibles": Array [
    "x",
    "light",
  ],
}
`;

exports[`ast gets test declarations with array 1`] = `
Array [
  Object {
    "at": 0,
    "constant": Object {
      "attributes": undefined,
      "identifiers": Array [],
      "name": "QUAD",
      "type": Object {
        "args": Array [
          Object {
            "args": Array [
              Object {
                "args": undefined,
                "name": "i32",
              },
            ],
            "name": "vec2",
          },
          Object {
            "name": "4",
          },
        ],
        "name": "array",
      },
      "value": "array<vec2<i32>, 4>(
        vec2<i32>(0, 0),
        vec2<i32>(1, 0),
        vec2<i32>(0, 1),
        vec2<i32>(1, 1),
			)",
    },
    "flags": 0,
    "symbol": "QUAD",
  },
]
`;

exports[`ast gets test empty function declaration 1`] = `
Array [
  Object {
    "at": 4,
    "flags": 0,
    "func": Object {
      "attributes": Array [
        Object {
          "args": undefined,
          "name": "export",
        },
      ],
      "identifiers": Array [],
      "name": "main",
      "parameters": undefined,
      "type": Object {
        "attributes": undefined,
        "name": "void",
      },
    },
    "symbol": "main",
  },
]
`;

exports[`ast gets test function declaration 1`] = `
Array [
  Object {
    "at": 4,
    "flags": 0,
    "func": Object {
      "attributes": Array [
        Object {
          "args": Array [
            "fragment",
          ],
          "name": "stage",
        },
      ],
      "identifiers": Array [
        "foo",
      ],
      "name": "fragShader",
      "parameters": Array [
        Object {
          "attributes": undefined,
          "name": "in1",
          "type": Object {
            "args": undefined,
            "name": "A",
          },
        },
        Object {
          "attributes": Array [
            Object {
              "args": Array [
                "2",
              ],
              "name": "location",
            },
          ],
          "name": "in2",
          "type": Object {
            "args": undefined,
            "name": "f32",
          },
        },
      ],
      "type": Object {
        "args": Array [
          Object {
            "args": undefined,
            "name": "f32",
          },
        ],
        "attributes": Array [
          Object {
            "args": Array [
              "0",
            ],
            "name": "location",
          },
        ],
        "name": "vec4",
      },
    },
    "symbol": "fragShader",
  },
]
`;

exports[`ast gets test imports 1`] = `
Array [
  Object {
    "at": 0,
    "imports": Array [
      Object {
        "imported": "MeshVertex",
        "name": "MeshVertex",
      },
    ],
    "name": "use/types",
    "symbols": Array [
      "MeshVertex",
    ],
  },
  Object {
    "at": 0,
    "imports": Array [
      Object {
        "imported": "viewUniforms",
        "name": "view",
      },
      Object {
        "imported": "worldToClip",
        "name": "worldToClip",
      },
    ],
    "name": "use/view",
    "symbols": Array [
      "view",
      "worldToClip",
    ],
  },
  Object {
    "at": 0,
    "imports": Array [
      Object {
        "imported": "getQuadUV",
        "name": "getQuadUV",
      },
    ],
    "name": "geometry/quad",
    "symbols": Array [
      "getQuadUV",
    ],
  },
  Object {
    "at": 0,
    "imports": Array [],
    "name": "test",
    "symbols": Array [],
  },
]
`;

exports[`ast gets test struct declaration 1`] = `
Array [
  Object {
    "at": 0,
    "flags": 0,
    "struct": Object {
      "attributes": undefined,
      "members": Array [
        Object {
          "attributes": undefined,
          "name": "intensity",
          "type": Object {
            "args": undefined,
            "name": "f32",
          },
        },
        Object {
          "attributes": Array [
            Object {
              "args": undefined,
              "name": "annotate",
            },
          ],
          "name": "position",
          "type": Object {
            "args": Array [
              Object {
                "args": undefined,
                "name": "f32",
              },
            ],
            "name": "vec3",
          },
        },
      ],
      "name": "light",
    },
    "symbol": "light",
  },
]
`;

exports[`ast gets test var/const declarations 1`] = `
Array [
  Object {
    "at": 0,
    "flags": 0,
    "symbol": "x",
    "variable": Object {
      "attributes": undefined,
      "identifiers": Array [],
      "name": "x",
      "qual": undefined,
      "type": Object {
        "args": undefined,
        "name": "f32",
      },
      "value": undefined,
    },
  },
  Object {
    "at": 18,
    "flags": 0,
    "symbol": "y",
    "variable": Object {
      "attributes": undefined,
      "identifiers": Array [],
      "name": "y",
      "qual": undefined,
      "type": Object {
        "args": undefined,
        "name": "f32",
      },
      "value": undefined,
    },
  },
  Object {
    "at": 36,
    "constant": Object {
      "attributes": undefined,
      "identifiers": Array [],
      "name": "a",
      "type": Object {
        "args": undefined,
        "name": "i32",
      },
      "value": "3",
    },
    "flags": 0,
    "symbol": "a",
  },
  Object {
    "alias": Object {
      "attributes": undefined,
      "name": "integer",
      "type": Object {
        "args": undefined,
        "name": "i32",
      },
    },
    "at": 58,
    "flags": 0,
    "symbol": "integer",
  },
  Object {
    "at": 81,
    "constant": Object {
      "attributes": undefined,
      "identifiers": Array [],
      "name": "b",
      "type": Object {
        "args": undefined,
        "name": "i32",
      },
      "value": undefined,
    },
    "flags": 0,
    "symbol": "b",
  },
]
`;
